### ISSUE TITLE
support keepassxc browser

### ISSUE THREAD

**Issue by Akruidenberg** [keeweb/keeweb#874]
It would be nice to support Keepassxc browser into the future.  Keepasshttp is less secure. an altrnative would be: https://github.com/smorks/keepassnatmsg

**Comment by antelle** [comment]
Or keepass-rpc, as an alternative.

**Comment by jeythekey** [comment]
Hi @antelle is there already any progress on that issue? [keepasshttp-connector](https://github.com/smorks/keepasshttp-connector) is deprecated and unchanged since almost one year.

If not, just an idea: A workaround could be to let keepassxc run in parallel and point it to the local/cached kdbx-file of keeweb. I tried it out and it works in principle, but external changes to the kdbx-file are apparently not recognized as such and hence not synced to the (webDAV) server by keeweb (they are, however, _displayed_ in keeweb immediately, but lost once I reload the file). I even cannot trigger a sync manually (with the menu button or by hitting ctrl+s).

So, would that be an easy fix to make keeweb aware of _any_ differences between the local and the remote file in order to trigger a syncing operation at the next occasion? That would be useful also in other scenarios, e.g. when using keepassxc for some features (such as removing fields) not (yet) implemented in keeweb.

Thanks for the great work by the way!

**Comment by antelle** [comment]
@jeythekey the described workaround won't work for the majority of users. We should better make the plugin support native messaging. For those who would like to contribute, the plugin is here: https://github.com/keeweb/keeweb-plugins/blob/master/docs/plugins/keewebhttp/plugin.js

**Comment by jeythekey** [comment]
Sure, that would be the ideal. Does your reply imply that the proposed syncing-fix is more involving? Because it would be nice also for other reasons. Thanks anyway for the quick reply!

**Comment by antelle** [comment]
The proposed sync doesn't seem to be real. What we consider sync, is when the file is changed by you (or someone else) on the remote storage, then we load it, sync changes, and upload back when you make other changes. The case when you do something on the same computer with another app (which doesn't want to sync it itself for some reason) while the file is still open in keeweb, and want to sync it immediately without changes, is not the real sync.

**Comment by jeythekey** [comment]
I understand, thanks. It was just unexpected, because the changes to the file are immediately visible in keeweb, so the local file known to keeweb _is_ different from the remote one, but keeweb doesn't know about that difference. Anyway, I understand this is probably something better not to loose time on, sorry for the noise.

**Comment by amamabel** [comment]
+1 for a Keepassxc browser support :-)

**Comment by dmuiX** [comment]
+1 for Keepassxc browser support

**Comment by antelle** [comment]
Since people here may find it interesting, I'll post my thoughts about supporting browser extensions in KeeWeb.

**KeePassHttp** or **KeeWebHttp** seems to be going nowhere, supporting it doesn't make sense anymore. There's a lot of critique about its security and I see that the browser extensions are not actively developed.

**KeePassXC-browser** is not bad, but:
1. There's an ethical dilemma regarding overwriting the application it opens, I would prefer describing it in the app but leave the choice to the user, so that this option is not offered by default but there has a possibility to do so, with a bit of configuration.
2. It doesn't work with Safari and the maintainers don't seem to be enthusiastic about it.

The protocol used in KeePassXC-browser called [**keepassxc-protocol**](https://github.com/keepassxreboot/keepassxc-browser/blob/develop/keepassxc-protocol.md) can be used by different extensions, so we should support it in KeeWeb. It will give us a possibility to connect to KeePassXC-browser, as well as build our own extension based on it.

**Own extension** doesn't seem to be hard to implement, assuming that we can avoid overloading it with features. The extension can provide the bare minimum that can be expected from it:
- insert username / password / both / auto / otp / other field
- submit the same as above
- trigger the auto-type selection pop-up if a matching entry can't be found automatically
- configurable shortcuts for common actions

What is not doable:
- auto-fill without user interaction: this is a product in itself that requires dedication and careful development due to vulnerabilities it can introduce and flexibility of HTML forms
- saving fields back to KeeWeb: again the absence of standards and APIs makes it complicated
- tooltips, icons, different auto-completion, etc... any other UI "injected" to websites: just don't do it! it looks and behaves not as it should

For example, we can start with something like this:

<img width="462" alt="Screen Shot 2021-03-22 at 21 47 28" src="https://user-images.githubusercontent.com/633557/112056405-3fe09d80-8b58-11eb-9f68-9573f40d860d.png">

What do you think? Is there demand for such extension?

Would this extension solve your day-to-day needs or something is missing?

**Comment by synchrone** [comment]
Thank you for considering this issue. Browser integration is a core feature, and I wouldn't be able to use a password manager without one.

First of all, KeepassXC-browser seems to be based on top those originally doing KeePassHttp, and it's a sizeable codebase for the seemingly simple feature.

To keep it light, considering backend, encryption and NativeMessaging proxy process still need to be implemented, and since the extension itself seems popular (150k users+) and working okay, why not just fork it, change NativeMessaging app request to KeeWeb, add credit, and republish to browser extension stores under the KeeWeb title to avoid confusion? 

This way it'd be easy to backport upstream fixes, security patches, and contribute efforts back inside the truly-free-password-management cohort

**Comment by antelle** [comment]
I'm not sure if forking it would be best in our case:

1. it's GPL, which means I may not publish an extension for Safari - that's I would say a blocker
2. the code needs to be tweaked a bit to clean up strange features from it, which will be harder
3. it still sounds not cool that the work is just forked and relicensed, even though the license allows this

**Comment by synchrone** [comment]
@antelle Forgive my ignorance, I had no idea Apple requires you to accept licensing that is not compatible with GPL in order to publish a Safari extension.

I would not propose to violate GPL of course.

In this case, since all of the original work (chromeIPass, etc) is GPL - you'd have to re-implement the browser extension from scratch under a more premissive license for no other reason than Apple's restrictions :(

**Comment by antelle** [comment]
Well, most of work in the extensions seems to be concentrated around these topics:
1. detecting login fields
2. auto-inserting username and password into the page
3. saving data back to KeeWeb
4. additional UI like icons and auto-completes
5. communication with the host

For now my vision that we'll implement a good alternative for auto-type, so only 5. will be required, which shouldn't be a lot of work.

**Comment by antelle** [comment]
I'd like to share some decisions and progress:

- we'll for sure have our own extension developed from scratch
- it will use KeePassXC-browser protocol with backward-compatible additions:
  - KeeWeb will be able to work with both KeePassXC-browser, as well as our own extension
  - there will be an option to switch to KeePassXC-browser in KeeWeb, but by default it will offer our extension
  - however, the extension won't be able to connect to KeePassXC (in theory, that's possible, but I don't think there's any demand for it)
- the UI will be minimalistic, think of it as auto-type for browser done right
- compared to KeePassXC-browser, this extension will provide:
  - support for web mode, when KeeWeb is opened in a browser tab, without installed native app
  - Safari support
  - better security, for example, it won't store database tokens in a plaintext file
  - seamless work with several databases being opened at the same time

The extension is developed here: https://github.com/keeweb/keeweb-browser
For now it can't do much, however it already can:
- connect to a KeeWeb tab in browser and exchange encrypted data
- use KeePassXC as backend

I hope to finish the work in v1.18, but as always, no guarantees.

Gray area:
- database connection token storage security: there are different solutions, I'm exploring them, it will most likely vary per browser, any input would be very welcome
- extension identity verification in the native app ("who connected to us?")

**Comment by gynet** [comment]
I really like the feature "connect to a KeeWeb tab in browser and exchange encrypted data", which is the one I was about to ask as a "Feature Request", you read my mind

**Comment by antelle** [comment]
🎉  First version of the browser extension is complete, the following should work:
- KeeWeb Connect <> KeeWeb tab: Chrome, Firefox, Edge
- KeeWeb Connect <> KeeWeb desktop app: Chrome, Firefox, Edge, Safari
- KeePassXC-Browser <> KeeWeb desktop tab: Chrome, Firefox, Edge

🐞  Most likely  there are a lot of bugs, but if anyone would like to give it a test, it's here: 
- [KeeWeb Connect](http://github.com/keeweb/keeweb-connect/releases/latest): only manual installation for now
- [KeeWeb desktop](https://github.com/keeweb/keeweb-ci-sandbox/releases/latest), or set the url to https://beta.keeweb.info/  to try a web-only mode

⛔️  What doesn't work:
- KeeWeb Connect <> KeeWeb tab in Safari: this is unlikely to happen, but the app should work
- KeeWeb Connect <> KeePassXC: won't be supported

🏃  TODO:
- fetching TOTP from KeeWeb: not implemented yet
- inserting other fields from KeeWeb in the browser extension

**Comment by VjokerA** [comment]
Please replace the icon with a more modern one.

**Comment by antelle** [comment]
Which icon?

**Comment by VjokerA** [comment]
Main, keeweb

**Comment by antelle** [comment]
You mean the extension? Which browser are you using? The icon should be up-to-date.

**Comment by VjokerA** [comment]
![image](https://user-images.githubusercontent.com/2880975/116686892-1ddf0400-a9bd-11eb-8b75-bd17c70b371d.png)

I think that the icon should be in a flat design, it would look fresh

**Comment by antelle** [comment]
I see, I thought something was wrong with the extension icon.
About this - nope 🤣  flat is not trendy anymore.

**Comment by gynet** [comment]
@antelle  I am testing it with Firefox + desktop(lastest + beta), seems conenction does not estiablish
Any extra settings in order to test the extension?

![image](https://user-images.githubusercontent.com/2174522/117494436-2ab5a780-af29-11eb-8c50-85bc4a112ccb.png)

**Comment by antelle** [comment]
What does it say in the background page of the extension?

**Comment by gynet** [comment]
Not see any thing on the extension option page's dev console, but on KeeWeb destop client saw this 
![image](https://user-images.githubusercontent.com/2174522/117495138-19b96600-af2a-11eb-9f94-7135129bf79e.png)

**Comment by gynet** [comment]
@antelle 
![image](https://user-images.githubusercontent.com/2174522/117495255-3eadd900-af2a-11eb-9798-a16feeea3629.png)

**Comment by antelle** [comment]
Not on the options page, the "background page" is a hidden page that can be inspected from chrome://extensions.

About the error in KeeWeb, is this the version downloaded from `keeweb-ci-sandbox`?

**Comment by gynet** [comment]
@antelle  I am using firefox.   No, the link you provide above was the production one, 1.17.6. Could you give me keeweb-ci-sandbox link so I can test it out?

**Comment by antelle** [comment]
Oops, I see my mistake, I posted the link in the post above, but probably it's a bit hidden. Here it is: https://github.com/keeweb/keeweb-ci-sandbox/releases/latest
The beta won't work because we also need the native part, and beta loads only the HTML. So, as expected, you can see it, but not use 😋

**Comment by gynet** [comment]
I just tested it on Firefox, it works like a charm, thanks!

**Comment by antelle** [comment]
Great, thanks! I think I'll make a release with it soon. The firefox extension is actually available in the Firefox store as well, you can use it from there directly.

**Comment by antelle** [comment]
Meanwhile, sharing some progress here. The extension is finished and is waiting for approval in the Edge store for the latest version (c’mon Microsoft, a week? even Apple approves Safari extensions in less than a day).
I’ll test it extensively this weekend and will release KeeWeb with it soon.

**Comment by antelle** [comment]
I'm pleased to let you know that support for both KeeWeb Connect and KeePassXC-Browser is merged into master and included in KeeWeb v1.18! 🎉 
KeeWeb Connect extension is now available in all browser stores and should work with the latest KeeWeb. This extension is the recommended way of browser integration, it works better with KeeWeb because it's developed with features of KeeWeb in mind.

You can give it a test and if you find anything strange, feel free to open a new issue!
Please be kind to the extension 🙏 , it's the first attempt in KeeWeb, so for now it's most likely not as great as it could be, it needs some testing and polishing to be production-ready.

☝️ It's a good idea to backup your data before upgrading KeeWeb!
